src/table/use-column-widths.tsx

```diff
@@ -63,26 +66,71 @@ const WidthsContext = createContext<WidthsContext>({
 interface WidthProviderProps {
   visibleColumns: readonly ColumnWidthDefinition[];
   resizableColumns: boolean | undefined;
+  containerRef: React.RefObject<HTMLElement>;
   children: React.ReactNode;
 }
 
-export function ColumnWidthsProvider({ visibleColumns, resizableColumns, children }: WidthProviderProps) {
-  const visibleColumnsRef = useRef<(PropertyKey | undefined)[] | null>(null);
-  const [columnWidths, setColumnWidths] = useState<Record<PropertyKey, number>>({});
+export function ColumnWidthsProvider({ visibleColumns, resizableColumns, containerRef, children }: WidthProviderProps) {
+  const visibleColumnsRef = useRef<PropertyKey[] | null>(null);
+  const containerWidthRef = useRef(0);
+  const [columnWidths, setColumnWidths] = useState<null | Record<PropertyKey, number>>(null);
 
   const cellsRef = useRef<Record<PropertyKey, HTMLElement>>({});
+  const stickyCellsRef = useRef<Record<PropertyKey, HTMLElement>>({});
   const getCell = (columnId: PropertyKey): null | HTMLElement => cellsRef.current[columnId] ?? null;
-  const setCell = (columnId: PropertyKey, node: null | HTMLElement) => {
+  const setCell = (sticky: boolean, columnId: PropertyKey, node: null | HTMLElement) => {
+    const ref = sticky ? stickyCellsRef : cellsRef;
     if (node) {
-      cellsRef.current[columnId] = node;
+      ref.current[columnId] = node;
     } else {
-      delete cellsRef.current[columnId];
+      delete ref.current[columnId];
     }
   };
+  // Imperatively sets width style for a cell avoiding React state.
+  // This allows setting the style as soon container's size change is observed.
+  const setColumnWidthStyle = useStableCallback((column: ColumnWidthDefinition) => {
+    const cellElement = cellsRef.current[column.id];
+
+    if (resizableColumns && columnWidths) {
+      const isLastColumn = column.id === visibleColumns[visibleColumns.length - 1]?.id;
+      const totalWidth = visibleColumns.reduce((sum, { id }) => sum + (columnWidths[id] || DEFAULT_COLUMN_WIDTH), 0);
+      if (isLastColumn && containerWidthRef.current > totalWidth) {
+        setElementWidth(cellElement, 'width', 'auto');
+      } else {
+        setElementWidth(cellElement, 'width', columnWidths[column.id]);
+      }
+      setElementWidth(cellElement, 'minWidth', column?.minWidth);
+    } else {
+      setElementWidth(cellElement, 'width', column?.width);
+      setElementWidth(cellElement, 'minWidth', column?.minWidth);
+    }
+    if (!resizableColumns) {
+      setElementWidth(cellElement, 'maxWidth', column?.maxWidth);
+    }
```

Comment:
Multiple `setElementWidth` may cause extra reflows can't they?

Also, there is a little bit of asymmetry, not all widths are reset in all conditions. Are you sure there will be no stale widths after state update?

How about refactoring this code into separate parts

```js
const {minWidth, width, maxWidth} = getColumnStyles(...);
setElementStyle(cellElement, {minWidth, width, maxWidth})
```

This solves both problems and makes the code more readable

Reply:
Thanks, the refactoring makes sense. In that particular case it won't make the performance better because the widths are all set in the thead render anyways. Those are updated imperatively only when there is a mismatch.